File: 22-generate-parentheses/22-generate-parentheses.java
    class Solution {
    public List<String> generateParenthesis(int n)
    {
    List<String> result=new ArrayList<>();
    helper(result,"",n,n);
    return result;
    }
    private void helper(List<String> result,String current,int left,int right)
    {
    if(left<0 || left>right) return;

    if(left==0 && right==0)
    {
    result.add(current);
    return;
    }
    helper(result,current+"(",left-1,right);
    helper(result,current+")",left,right-1);
    }
    }

File: 22-generate-parentheses/22-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def helper(current,left,right,res):
            if(left<0 or left>right): return;
            if(left==0 and right==0):
                res.append(current)
                return
            helper(current+"(",left-1,right,res)
            helper(current+")",left,right-1,res)
        res=[]
        helper("",n,n,res)
        return res